### ISSUE TITLE
Callback on Request completion to construct Response

### ISSUE THREAD

**Issue by jerinphilip** [browsermt/bergamot-translator#65]
The changes proposed in this issue is to be implemented if approved following Alignments PR (#46), and can help #56, #53.

tl;dr: `f(x; \theta) = ResponseBuilder(histories; vocabs, std::promise<Response>)`

It's a bit unnatural to have `std::promise<Response>` and `vocabs` in `Request`, which should only hold request information. 

https://github.com/browsermt/bergamot-translator/blob/317433af9fe6e4f7ad56e41ab47d7dfbe5cdbabd/src/translator/request.h#L38-L40

Similarly `Response` is constructed from `Histories` and `Vocabs` (both of which needn't be there, #53). This can be simplified / neatened by having a `ResponseBuilder` outside `Response`, breaking the following messy constructor into meaningful units.

https://github.com/browsermt/bergamot-translator/blob/317433af9fe6e4f7ad56e41ab47d7dfbe5cdbabd/src/translator/response.cpp#L11-L103 

`ResponseBuilder` will handle taking in `histories` and be initialized with `vocabs` and the `promise`. Using `histories` and `vocab`, moving the present constructor of `Response` into `ResponseBuilder` can enable the construction of a `Response` there, following which the `std::promise<Response>` can be set with the newly constructed instance of `Response`. 

1. `Response` will thus end up carrying only data members (`AnnotatedBlob`s of source, target. `QualityScore`s, `Alignments`). 
2. Consequently, `Response` should become very thin, and in theory ready for WASM export (conditioned on `Annotation` being exportable). No `string_view` offending, `alignments` and stub `QualityScores` ready.
3. An instance of `ResponseBuilder` initialized with vocabs and promise and accepting histories additionally can be registered as a callback to `Request` instead of the existing spread mechanism (to be fired once translation of request is completed), consolidating the transition from processed `Request` -> `Response` logic into this callback.
4. `QualityEstimation` people can be pointed towards just `ResponseBuilder`, where they will have additional access to `histories` to just operate and vocabs etc. Feels like a saner API.
5. When amend/cancel capable futures are required, the `std::promise` can be replaced with `std::promise` equivalent of the enhanced future.
